velox/functions/lib/SubscriptUtil.h

```diff
@@ -96,6 +97,42 @@ class LookupTable : public LookupTableBase {
   std::unique_ptr<outer_map_t> map_;
 };
 
+// A flat vector of map keys, an index into that vector and an index into
+// the original map keys vector that may have encodings.
+struct MapKey {
+  const BaseVector* baseVector;
+  const vector_size_t baseIndex;
+  const vector_size_t index;
+
+  size_t hash() const {
+    return baseVector->hashValueAt(baseIndex);
+  }
+
+  bool operator==(const MapKey& other) const {
+    return baseVector->equalValueAt(
+        other.baseVector, baseIndex, other.baseIndex);
+  }
+
+  bool operator<(const MapKey& other) const {
+    return baseVector->compare(other.baseVector, baseIndex, other.baseIndex) <
+        0;
+  }
+};
+
+struct MapKeyHasher {
+  size_t operator()(const MapKey& key) const {
+    return key.hash();
+  }
+};
+
+using MapKeyAllocator = memory::StlAllocator<detail::MapKey>;
+
+using ComplexKeyHashMap = folly::F14FastSet<
```

Comment:
Can we change to use folly::F14FastMap instead?

Reply:
It could be a Map too, I'm not sure if keeping everything in the Key is better for memory locality or reducing overhead, but I also didn't see a strong reason to use a Map instead of a Set so I didn't want to risk changing it.